File: c++/day11.cpp
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <list>


std::list<unsigned long long> readInput() {
  std::list<unsigned long long> result;
  std::ifstream inputStream("../inputs/11.txt");
  std::string line;
  unsigned long long rowIndex = 0;

  while (std::getline(inputStream, line)) {
    std::stringstream ss(line);
    unsigned long long number;
    while(ss >> number) {
      result.push_back(number);
    }
  }

  inputStream.close();
  return result;
}

int getDigitCount(unsigned long long n) {
  if(n == 0){
    return 1;
  }
  return (unsigned long long)std::log10(n) + 1;
}

std::pair<unsigned long long, unsigned long long> splitNumberInHalf(unsigned long long n, unsigned long long digits) {
  unsigned long long half = digits / 2;
  unsigned long long split = std::pow(10, half);

  return {n / split, n % split};
}


void singleBlink (std::list<unsigned long long>& stones) {
  for (auto it = stones.begin(); it != stones.end(); ++it){
    unsigned long long stone = *it;
    if(stone == 0) {
      *it += 1;
    }
    else if (unsigned long long digits = getDigitCount(stone); digits % 2 ==0){
      std::pair<unsigned long long, unsigned long long> split = splitNumberInHalf(stone, digits);
      stones.insert(it, split.first);
      *it = split.second;
    }
    else {
      *it *= 2024;
    }
  }
}


unsigned long long simulateBlinks(std::list<unsigned long long> stones, unsigned long long blinks) {
  std::list<unsigned long long> zeroed = {0};
  for(unsigned long long i = 0 ; i < blinks; i++) {
    std::cout << "Blink " << i << "\n";
    std::cout<< stones.size() << "\n\n";
    singleBlink(zeroed);
  }
  int zeroSubstitute = zeroed.size();
  unsigned long long zerosSum = 0;

  for(auto it = stones.begin(); it != stones.end(); ++it) {
    if(*it == 0) {
      zerosSum += zeroSubstitute;
      it = stones.erase(it);
    }
  }


  for(unsigned long long i = 0 ; i < blinks; i++) {
    std::cout << "Blink " << i << "\n";
    std::cout<< stones.size() << "\n\n";
    singleBlink(stones);
  }

  return stones.size() + zerosSum;
}




unsigned long long silver() { 
  auto input = readInput();
  return simulateBlinks(input, 25);
}

unsigned long long gold() {
  auto input = readInput();
  return simulateBlinks(input, 75);
}

int main() {
  std::cout << "silver: " << silver() << "\n";
  std::cout << "gold: " << gold() << "\n";
}

File: python/day11.py
from functools import lru_cache
import math


def read_input(filename):
    with open(filename, "r") as f:
        return [int(n) for n in f.read().split(" ")]


def get_digit_count(n):
    return 1 if n == 0 else math.floor(math.log10(n)) + 1


def split_number(n, digits):
    half = digits // 2
    split = 10**half
    return (n // split, n % split)


@lru_cache(maxsize=None)
def simulate_blink(stone, remaining):
    count = 0
    if remaining == 0:
        return 1

    if stone == 0:
        count += simulate_blink(1, remaining - 1)
    elif (digits := get_digit_count(stone)) % 2 == 0:
        splitted = split_number(stone, digits)
        count += simulate_blink(splitted[0], remaining - 1)
        count += simulate_blink(splitted[1], remaining - 1)
    else:
        count += simulate_blink(stone * 2024, remaining - 1)

    return count


def silver(input):
    return sum([simulate_blink(stone, 25) for stone in input])


def gold(input):
    return sum([simulate_blink(stone, 75) for stone in input])


if __name__ == "__main__":
    input = read_input("../inputs/11.txt")
    print(f"silver {silver(input)}")
    print(f"gold {gold(input)}")
